File: code/graphs/union-find.cpp
// Weighted quick-union with path compression. Two arrays: parent[i] (who i
// points to) and size[i] (tree size, for union by size). Near-constant per op.
struct UF {
    vector<int> parent, size;
    UF(int n) : parent(n), size(n, 1) {
        for (int i = 0; i < n; i++) parent[i] = i;   // each element its own root
    }
    int find(int x) {                        // walk to the root, compressing
        while (x != parent[x]) {
            parent[x] = parent[parent[x]];    // path halving: point at grandparent
            x = parent[x];
        }
        return x;
    }
    void unite(int a, int b) {               // union by size: smaller under larger
        int ra = find(a), rb = find(b);
        if (ra == rb) return;
        if (size[ra] < size[rb]) swap(ra, rb);
        parent[rb] = ra;                      // rb (smaller) hangs under ra
        size[ra] += size[rb];
    }
    bool connected(int a, int b) { return find(a) == find(b); }
};

File: code/graphs/union-find.rs
// Weighted quick-union with path compression. parent[i] = i's parent,
// size[i] = tree size (for union by size). Near-constant amortized per op.
struct UF {
    parent: Vec<usize>,
    size: Vec<usize>,
}
impl UF {
    fn new(n: usize) -> Self {
        UF { parent: (0..n).collect(), size: vec![1; n] }
    }
    fn find(&mut self, mut x: usize) -> usize {   // walk up, compressing
        while x != self.parent[x] {
            self.parent[x] = self.parent[self.parent[x]];  // path halving
            x = self.parent[x];
        }
        x
    }
    fn unite(&mut self, a: usize, b: usize) {     // union by size
        let (mut ra, mut rb) = (self.find(a), self.find(b));
        if ra == rb { return; }
        if self.size[ra] < self.size[rb] {
            std::mem::swap(&mut ra, &mut rb);          // smaller under larger
        }
        self.parent[rb] = ra;
        self.size[ra] += self.size[rb];
    }
    fn connected(&mut self, a: usize, b: usize) -> bool {
        self.find(a) == self.find(b)
    }
}
